### PR TITLE
Allow mutating queue name in StatefulSet Webhook.

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Allow mutating queue name in StatefulSet Webhook.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Allow mutating queue name in StatefulSet Webhook.
```

### CODE REVIEW

**Comment by mimowo** [comment]
/hold
I want to understand the flow e2e first from the user perspective.
In particular, how can user start such a StafulSet, will adding the label make it start? 

IIRC for Jobs we start such a Job (but please double-check and confirm).

I synced with @mbobrovskyi that this is to align the behavior for Deployment, but another option is to simply reject such Deployments if they are not supported anyway.

I think it deserves e2e test.

**Comment by mimowo** [comment]
> In the Pod webhook Default method, if the namespaceSelector doesn't match then we never get to the code that consults manageJobsWithoutQueueName. 
 
 Yes, this is WAI. The intention was to have a mechanism to exclude pods (like static pods or DeamonSet pods) in kube-system and kueue-system. We made the mechanism more generic (to exclude arbitrary namespaces). 
 
> We don't support namespaceSelectors to modify manageJobsWithoutQueueName for any other integration (discussed at length in https://github.com/kubernetes-sigs/kueue/issues/2119).

Right, we don't do it for all other integrations. However, I think Deployments and StatefulSets need to be the other cases, first Deployments are used in kube-system and kueue-system so we better don't touch them. Second, the support is based on the PodGroup integration and so we inherit the lookup into `namespaceSelector` for the pod integration.

> What is the intended semantics for a StatefulSet or Deployment that is deployed in a namespace that doesn't match the namespaceSelector for the Pod integration when manageJobsWithoutQueueName is true?

IIUC this means basically "for Deployments and StatefulSets in the kube-system or kueue-system". I think we should not manage them - no workload should be created. Since Deployments and StatefulSets are based on PodGroup integration this should happen "for free".

Let me know if this matches your expectations and understanding. 

cc @mwielgus

**Comment by dgrove-oss** [comment]
It honestly feels a bit like our implementation is leaking through to the API.   In particular, treating StatefulSets one way and Jobs another wrt manageJobsWithoutQueueName.

I think it could be less surprising / easier to explain if the boolean `manageJobsWithoutQueueNames` was replaced with a namespaceSelector across all integrations.   I know this was discussed before, but maybe it is worth revisiting now that (a) we see what we need for Deployment and StatefulSet and (b) we are thinking about what a v1 API would look like and what perhaps should be improved between now and then.

**Comment by mimowo** [comment]
> It honestly feels a bit like our implementation is leaking through to the API. In particular, treating StatefulSets one way and Jobs another wrt manageJobsWithoutQueueName.

Yeah, I see the point - so that it is not clear why StatefulSet or Deployment pods are controlled by `namespaceSelector`, whilst for other Jobs this is not respected.

> I think it could be less surprising / easier to explain if the boolean manageJobsWithoutQueueNames was replaced with a namespaceSelector across all integrations.

You mean "replaced"? Or something like "restricted" - so that we only manage workloads matching the namespaceSelector?

> I know this was discussed before, but maybe it is worth revisiting now that (a) we see what we need for Deployment and StatefulSet and (b) we are thinking about what a v1 API would look like and what perhaps should be improved between now and then.

I would be in favor of that. The original intention of podOptions.namespaceSelector was to exclude "kube-system" and "kueue-system" from pods. Back then we didn't foresee the need to exclude managing for Jobs or other supported CRDs. However, as we now support Deployments it makes also sense to exclude "kube-system" and "kueue-system". Luckily this is for free by using Pod integration, but as you say it means leaking implementation details.

Let me also cc @mwielgus and @tenzen-y for their opinions, but +1 from me to decouple `namespaceSelector` from `podOptions`. 

The remaining question from me: do we support both places, or we validate only one is set? We could consider supporting both places for v1beta1 and depracate the one in podOptions, but it would be good to have a KEP for that. Are you interested in driving this?

**Comment by dgrove-oss** [comment]
> You mean "replaced"? Or something like "restricted" - so that we only manage workloads matching the namespaceSelector?

restricted is a better word :).

Yes, I'd propose that we do a uniform filtering by namespaceSelector for all integrations when manageJobsWithoutQueueName is true.   I'll give people some time to comment, but if there is interest in exploring this I'd be happy to kick off a KEP and drive it.

File: pkg/controller/jobs/statefulset/statefulset_controller.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ const (
 func init() {
 	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
 		SetupIndexes:   SetupIndexes,
-		NewReconciler:  NewReconciler,
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_controller.go:42]: Why is this changed? Please don't if not necessary.

File: pkg/controller/jobs/statefulset/statefulset_webhook.go [hunk 2]
```diff
@@ -79,13 +80,17 @@ func (wh *Webhook) Default(ctx context.Context, obj runtime.Object) error {
...
-			ss.Spec.Template.Labels = make(map[string]string, 2)
-		}
 		queueName := jobframework.QueueNameForObject(ss.Object())
 		if queueName != "" {
+			if ss.Spec.Template.Labels == nil {
+				ss.Spec.Template.Labels = make(map[string]string, 2)
+			}
 			ss.Spec.Template.Labels[constants.QueueLabel] = queueName
-			ss.Spec.Template.Labels[pod.GroupNameLabel] = GetWorkloadName(ss.Name)
+			groupName, err := GetWorkloadName(obj.(*appsv1.StatefulSet))
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_webhook.go:89]: This change does not seem releated to changing queue-name. Please revert if not necessary.

**Review by mimowo — COMMENTED** [review]
This PR seems to be changing much more than needed. Let me hold until the scope and purpose of the PR is clarified.
/hold

File: pkg/controller/jobs/statefulset/statefulset_controller.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/jobs/statefulset/statefulset_controller.go:42]: I don't understand this case - can you elaborate? For example in the core k8s Job we also have finalizers on Job pods, but the reconciler is at the Job level, [reference](https://github.com/kubernetes/kubernetes/blob/f1b3fdf7e6d40714b1a43757221832aa1c4a49d1/pkg/controller/job/job_controller.go#L1173C23-L1173C56). I would like to first document such problematic scenarios in some form of tests and change the implementation in a dedicated PR (if needed).

File: pkg/controller/jobs/statefulset/statefulset_webhook.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_webhook.go:89]: Agree. Created separate PR #3951.

File: pkg/controller/jobs/statefulset/statefulset_controller.go (hunk 1 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_controller.go]: > I don't understand this case - can you elaborate?

After implementing scale-up/scale-down for StatefulSets, when I tried to forcefully change the replicas, I found an issue where the finalizer was not deleted from the Pod. This can occur because we have a reconciler for the StatefulSet but not for the Pod. 

> For example in the core k8s Job we also have finalizers on Job pods, but the reconciler is at the Job level, [reference](https://github.com/kubernetes/kubernetes/blob/f1b3fdf7e6d40714b1a43757221832aa1c4a49d1/pkg/controller/job/job_controller.go#L1173C23-L1173C56).

To ensure the phase of the Pod changes to Failed or Succeeded, it’s better to reconcile the Pod directly instead of fetching the list of Pods for each change on the StatefulSet. Additionally, we don’t need a reconciler for the StatefulSet at all.
Alternatively, we would need to reconcile the StatefulSet for each change to each Pod and fetch the list of Pods, which could increase the load on the database.

> I would like to first document such problematic scenarios in some form of tests and change the implementation in a dedicated PR (if needed).

These changes are not related to queue-name changes, so I propose continuing the discussion in https://github.com/kubernetes-sigs/kueue/pull/3951. The issue occurs in [this test case](https://github.com/kubernetes-sigs/kueue/blob/3a089be14a30552aae8b18a3b43bd460d82a6380/test/e2e/singlecluster/statefulset_test.go#L319).

Comment (mimowo) [reply pkg/controller/jobs/statefulset/statefulset_controller.go]: The PR for arbitrary scale-down and up looks quite complex, and there is not much demand from users for it at the moment, so I'm deprioritizing the review of it. 

is there a way to support mutating the "queue-name" without it? It isn't clear to me why the things are coupled.

Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_controller.go]: I moved this logic to separate #3951 PR. So this PR just allow mutating queue name in StatefulSet Webhook.

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 3]
```diff
@@ -56,42 +57,65 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+		}))
 	})
 }
 
+func finalizePod(sts *appsv1.StatefulSet, pod *corev1.Pod) bool {
+	var updated bool
+
+	if shouldUngate(sts, pod) && utilpod.Ungate(pod, podcontroller.SchedulingGateName) {
+		updated = true
+	}
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:101]: FYI: When a Pod has a scheduling gate, it should be ungated because StatefulSet does not allow updating Pods with a scheduling gate.

Comment (mimowo) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go:101]: sgtm

File: test/e2e/singlecluster/statefulset_test.go [hunk 4]
```diff
@@ -325,5 +328,49 @@ var _ = ginkgo.Describe("StatefulSet integration", func() {
 				}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
 			})
 		})
+
+		ginkgo.It("should allow to change queue name if ReadyReplicas=0", func() {
+			statefulSet := statefulsettesting.MakeStatefulSet("sts", ns.Name).
+				Image(util.E2eTestSleepImage, []string{"10m"}).
+				Request(corev1.ResourceCPU, "100m").
+				Replicas(3).
+				Queue(fmt.Sprintf("%s-invalid", localQueueName)).
```
Comment (mimowo) [inline test/e2e/singlecluster/statefulset_test.go:337]: I think the more practical case for updating is setting if by users who forgot to set it in the first place.
So, I would suggest to remove it and set to the correct value. We could have both e2e tests, but maybe this is overkill.

File: pkg/controller/jobs/statefulset/statefulset_webhook_test.go [hunk 5]
```diff
@@ -249,76 +249,39 @@ func TestValidateUpdate(t *testing.T) {
 			wantErr: nil,
 		},
 		"change in queue label": {
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_webhook_test.go:251]: Add a test case for setting the queue label for the first time.

**Review by mimowo — COMMENTED** [review]
Thanks!
LGTM, just minor suggestions for the testing coverage

File: test/e2e/singlecluster/statefulset_test.go (hunk 4 above)
Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go:337]: If a user forgets to set the queue name, the Pods should run. Is it valid case?

Comment (mimowo) [reply test/e2e/singlecluster/statefulset_test.go:337]: Hmmm, I was thinking about scenario when using `manageJobsWithoutQueueName: true`. 
However, since our testing assumes `manageJobsWithoutQueueName: false` then the test case is correct. 
We can extend the testing after https://github.com/kubernetes-sigs/kueue/issues/3767, cc @kaisoz .

So feel free to keep as is.

Comment (mimowo) [reply test/e2e/singlecluster/statefulset_test.go:337]: In case the Pods are already running, and we have the `manageJobsWithoutQueueName: false`, ideally we stop the STS and restart as managed by Kueue, but I'm not sure what is the priority, because no user raised this with us yet. They can always in all cases just re-create it. So, I'm not worried about the case yet. 

However, if you test that it does not work as expected, then it is better to prevent it by validation.

Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go:337]: We discussed it offline with @mimowo and decided to keep it as is for now.

File: pkg/controller/jobs/statefulset/statefulset_webhook_test.go (hunk 5 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_webhook_test.go:251]: Added on 972193d.

Comment (mimowo) [reply pkg/controller/jobs/statefulset/statefulset_webhook_test.go:251]: lgtm

**Comment by mimowo** [comment]
> /hold
> I want to understand the flow e2e first from the user perspective.
> In particular, how can user start such a StafulSet, will adding the label make it start?

/hold cancel
I believe this is covered now with the e2e test which demonstrates how a user can change the label and the StatefulSet will start to run.

**Comment by mimowo** [comment]
> It honestly feels a bit like our implementation is leaking through to the API. In particular, treating StatefulSets one way and Jobs another wrt manageJobsWithoutQueueName.

I believe this concern is already addressed by the KEP: https://github.com/kubernetes-sigs/kueue/issues/3589. We are yet to follow up on deprecating and removing the podOptions, but I wouldn't block this work on that.

**Review by mimowo — COMMENTED** [review]
/lgtm
/approve
